core/indigo-core/molecule/metadata_storage.h

```diff
@@ -0,0 +1,71 @@
+/****************************************************************************
+ * Copyright (C) from 2009 to Present EPAM Systems.
+ *
+ * This file is part of Indigo toolkit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+#ifndef __meta_data_storage__
+#define __meta_data_storage__
+#include "base_cpp/ptr_array.h"
+
+namespace indigo
+{
+    class MetaObject
+    {
+    public:
+        explicit MetaObject(uint32_t class_id) : _class_id(class_id)
+        {
+        }
+        uint32_t _class_id;
+        virtual MetaObject* clone() const = 0;
+        virtual ~MetaObject(){};
+    };
+
+    class MetaDataStorage
+    {
+    public:
+        void cloneMetaData(const MetaDataStorage& other);
+
+        virtual ~MetaDataStorage()
+        {
+        }
+
+        void addMetaObject(MetaObject* pobj);
+
+        void resetMetaData()
+        {
+            _meta_data.clear();
+            _plus_indexes.clear();
+            _arrow_indexes.clear();
```

Comment:
I believe metadata for molecule and reaction should be different — there should not be any `plus` or `arrow` in molecule metadata. Perhaps we should think about inheritance here?

Reply:
Metadata shouldn't be different for molecules and reactions as far as plus or arrow doesn't always mean "reaction".
It should be possible to load arrows and pluses with molecules.